projects/beta_49_ssl_handshake_version_spike.kp/knowledge.md

```diff
@@ -0,0 +1,246 @@
+---
+title: SSL_HANDSHAKE_VERSION Beta 49 Spike
+authors:
+- Frank Bertsch
+tags:
+- beta 49
+- spike
+created_at: 2017-04-22 00:00:00
+updated_at: 2017-05-22 12:21:42.436900
+tldr: A single client sent hundreds of thousands of duplicate pings, all with huge
+  counts of SSL_HANDSHAKE_VERSION
+thumbnail: images/output_9_1.png
+---
+# SSL_HANDSHAKE_VERSION Spike
+
+TL;DR: A single client sending duplicate pings with huge counts of SSL_HANDSHAKE_VERSION caused this spike.
+
+[See the prior work here](https://github.com/ekr/telemetry-repro/pull/1), where I determined that the number of Measurements of SSL_HANDSHAKE_VERSION starting on 20170308 for Beta 49 increased by more than more than 3 orders of magnitude.
+
+The associated ping count only increased by ~50%. This indicated that the problem was not just ping volume, but mostly histogram counts within the pings.
+
+## Looking at 20170312 - the worst day
+
+To find these offenders we'll look at the day with the highest count - $587 * 10^3$ pings but $113 * 10^9$ SSL_HANDSHAKE_VERSION measurements.
+
+
+```python
+from moztelemetry.dataset import Dataset
+from moztelemetry.spark import get_pings_properties
+
+from datetime import datetime, timedelta
+import pandas as pd
+import numpy as np
+```
+    Unable to parse whitelist: /mnt/anaconda2/lib/python2.7/site-packages/moztelemetry/histogram-whitelists.json.
+    Assuming all histograms are acceptable.
+
+
+
+```python
+pings = Dataset \
+.from_source('telemetry') \
+.where (
+    appUpdateChannel = 'beta',
+    appVersion = lambda x: x.startswith('49'),
+    submissionDate = '20170312'
+).records(sc)
+```
+    fetching 16579.50070MB in 6728 files...
+
+
+
+```python
+def get_shv(ping):
+    return ping.get('payload', {}).get('histograms', {}).get('SSL_HANDSHAKE_VERSION', {})
+
+per_client = \
+pings.map(
+    lambda x: 
+        (x.get('clientId'),
+         get_shv(x))
+).filter(
+    lambda x: x[0] and x[1]
+).map(
+    lambda x: (x[0], {'sum': x[1]['sum'], 'count': sum(x[1]['values'].values())})
+).reduceByKey(
+    lambda x, y: {k: x[k] + y[k] for k in x}
+)
+
+#How many clients?
+"Total Clients for Beta 49 on 20170312: {}".format(per_client.count())
+```
+
+
+
+
+    'Total Clients for Beta 49 on 20170312: 62423'
+
+
+
+
+```python
+#How bad is the worst offender?: > 100B Measurements
```

Comment:
This inline comment Q&A would be better after the fact as conclusions and analysis. Inline it can be missed.